oak_runtime/introspection_browser_client/index.tsx

```diff
@@ -45,6 +45,104 @@ function loadSerializedEvents(): Promise<Uint8Array> {
   });
 }
 
+interface OakApplicationState {
+  nodeInfos: NodeInfos;
+  channels: Channels;
+}
+
+type NodeId = number;
+type AbiHandle = number;
+type NodeInfos = Map<NodeId, NodeInfo>;
+interface NodeInfo {
+  abiHandles: Map<AbiHandle, ChannelHalf>;
+}
+
+type ChannelID = number;
+enum ChannelHalfDirection {
+  Read,
+  Write
+}
+interface ChannelHalf {
+  channelId: ChannelID;
+  direction: ChannelHalfDirection;
+}
+interface Message {
+  data: Uint8Array;
+  channels: ChannelHalf[];
+}
+interface Channel {
+  id: ChannelID;
+  messages: Message[];
+}
+type Channels = Map<ChannelID, Channel>;
+
+function eventReducer(
+  applicationState: OakApplicationState,
+  event: introspectionEventsProto.Event
+) {
+  const eventType = event.getEventDetailsCase();
+  const { EventDetailsCase } = introspectionEventsProto.Event;
+
+  switch (eventType) {
+    case EventDetailsCase.NODE_CREATED:
+      applicationState.nodeInfos.set(event.getNodeCreated().getNodeId(), {
+        abiHandles: new Map()
+      });
+
+      break;
+    case EventDetailsCase.NODE_DESTROYED:
+      applicationState.nodeInfos.delete(event.getNodeDestroyed().getNodeId());
+
+      break;
+    case EventDetailsCase.CHANNEL_CREATED:
+      {
+        const channelId = event.getChannelCreated().getChannelId();
+        applicationState.channels.set(channelId, {
+          id: channelId,
+          messages: []
+        });
+      }
+
+      break;
+    case EventDetailsCase.CHANNEL_DESTROYED:
+      applicationState.channels.delete(
+        event.getChannelDestroyed().getChannelId()
+      );
+
+      break;
+    case EventDetailsCase.HANDLE_CREATED:
+      {
+        const details = event.getHandleCreated();
+        applicationState.nodeInfos
+          .get(details.getNodeId())
+          .abiHandles.set(details.getHandle(), {
+            channelId: details.getChannelId(),
```

Comment:
Does it require a check to make sure the `channelId` exists in `applicationState.channels`?

Reply:
If introspection events are fired correctly the `channelId ` should always exist. In case it doesn't, a runtime error will occur, along the lines of `"Cannot read property 'abiHandles' of undefined"`. 

I could add a check here to include a more descriptive error message, though I'm not sure how much value it'd add. :)

Reply:
Turns out that I can enable a setting TypeScript so this does require a check. (`strictNullChecks`). Added a commit to enable that & add null checks with proper error messages, similar to how it is done in Rust.

Reply:
Nice. I was kind of thinking about how `nodeInfos` and `channels` in `OakApplicationState` remain in sync. If that is at all necessary. I guess, as you mentioned, if the events are fired correctly, `nodeInfos` and `channels` will be synced.